pkg/grpc/exec/exec.go

```diff
@@ -369,13 +366,12 @@ func (msg *MsgExitEventUnix) RetryInternal(ev notify.Event, timestamp uint64) (*
 	}
 
 	if internal != nil {
+		// Use cached version of the process
+		ev.SetProcess(internal.UnsafeGetProcess())
```

Comment:
Can you remind me why this is safe? We dec the reference after this, but create a reference to the pointer here? The point is that we are starting the cache removal here and hoping that the event leaves the system before the destroy happens?

Reply:
I think its the same logic used on HandleMessage() so should be ok?

Reply:
OK nevermind switching runtimes C <->Golang created some brain melt.

So basic idea here is Exit event is telling us the object can be removed assuming no child reference it with a refinc. After Exit we may get kprobe events if they are OOO but this is OK because we only kick of the removal process. There is going to be a delay betwen exit dec'ing the refcnt and its actual removal from the cache. Finally it also doesn't matter if its removed before the Exit event is pushed because golang runtime knows we have a reference to it here and wont free the actual memory.

Reply:
> OK nevermind switching runtimes C <->Golang created some brain melt.
> 
> So basic idea here is Exit event is telling us the object can be removed assuming no child reference it with a refinc. After Exit we may get kprobe events if they are OOO but this is OK because we only kick of the removal process. There is going to be a delay betwen exit dec'ing the refcnt and its actual removal from the cache. Finally it also doesn't matter if its removed before the Exit event is pushed because golang runtime knows we have a reference to it here and wont free the actual memory.

Indeed, so with the RefDec() if it is zero we push it to the deletequeue that is the delay between exit dec'ing the refcnt and the actual removal from the cache, the GC then will trigger at first round at X time for this process to be deleted which will set its color to deleteReady, then a second trigger after 30sec will then remove the process from the cache, during this time if an out of order events comes kprobes after we RefDec the process at exit, then the event cache triggers every 2 secs for a total of 15 cache strikes which covers the same duration, so I guess we are ok , and if by chance the process is removed and the event cache retry returns nil on the process then we have the metrics on the ProcessInfo or PodInfo errors to indicate that...

Great work btw ;-) !  and for C<->Golang and golang runtime couldn't say it better haha :-D !